File: 05-Hash Tables/java/src/CharFinder.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class CharFinder {

    public char findFirstNonRepeatingChar(String str) {
        Map<Character, Boolean> map = new HashMap<>();

        var chars = str.toCharArray();
        for (var ch : chars)
            map.put(ch, !map.containsKey(ch));

        for (var ch : chars)
            if (map.get(ch))
                return ch;

        return Character.MIN_VALUE;
    }

    public char findFirstRepeatedChar(String str) {
        Set<Character> set = new HashSet<Character>();

        for (var ch : str.toCharArray()) {
            if (set.contains(ch))
                return ch;

            set.add(ch);
        }

        return Character.MIN_VALUE;
    }

}

File: 05-Hash Tables/cpp/CharFinder.cpp
#include "CharFinder.h"
#include<unordered_map>
#include<set>

char CharFinder::findFirstNonRepeatingChar(string str) {
    unordered_map<char, int> unorderedMap;

    for (char ch: str) {
        if (0 == unorderedMap.count(ch))
            unorderedMap[ch] = 1;
        else
            unorderedMap[ch] += 1;
    }

    for (char ch: str) {
        if (1 == unorderedMap.at(ch))
            return ch;
    }


    return CHAR_MIN;
}

char CharFinder::findFirstRepeatedChar(string str) {
    set<int> s;

    for (char ch: str) {
        if (s.count(ch))
            return ch;

        s.insert(ch);
    }

    return CHAR_MIN;
}
